### ISSUE TITLE
Vert.x threads block after brief load testing

### ISSUE THREAD

**Issue by geoand** [quarkusio/quarkus#4173]
Using `master` (specifically with commit: 6fd65a24510282d6a04af360bedc28720188aa15) I am seeing that when doing a small load testing run with [this](https://github.com/geoand/quarkus-vertx-threads-blocked) very simple reproducer project, vertx threads ending up being blocked and the application becoming unresponsive.

The threads seem to be blocked here:

```
io.vertx.ext.web.handler.impl.StaticHandlerImpl.isFileExisting(RoutingContext, String, Handler) StaticHandlerImpl.java:275
```

I should note that this doesn't happen in the `0.23-wip` branch

**Comment by cescoffier** [comment]
@pmlopes wondering if we can removed the synchronized in https://github.com/vert-x3/vertx-web/blob/91d92726042f2694ac55749736d7411e8be7e0b2/vertx-web/src/main/java/io/vertx/ext/web/handler/impl/StaticHandlerImpl.java#L277

**Comment by cescoffier** [comment]
The contention may come from this. In Quarkus we have several event loops using the handler.

**Comment by johnaohara** [comment]
I ran the reproducer locally. After the app became unresponsive, the vertx worker threads were all in a WAIT state 

```
"vert.x-worker-thread-19" #63 prio=5 os_prio=0 tid=0x00007f7958028800 nid=0x6327 in Object.wait() [0x00007f794dbe9000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x00000006c7fdd7c8> (a io.vertx.core.http.impl.Http1xServerConnection)
        at java.lang.Object.wait(Object.java:502)
        at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.awaitWriteable(VertxBlockingOutput.java:96)
        at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.write(VertxBlockingOutput.java:63)
        - locked <0x00000006c7fdd7c8> (a io.vertx.core.http.impl.Http1xServerConnection)
        at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.writeBlocking(VertxHttpResponse.java:179)
        at io.quarkus.resteasy.runtime.standalone.VertxOutputStream.close(VertxOutputStream.java:118)
        at org.jboss.resteasy.util.CommitHeaderOutputStream.close(CommitHeaderOutputStream.java:87)
        at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.finish(VertxHttpResponse.java:136)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:109)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:64)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:55)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$$Lambda$139/2023668063.handle(Unknown Source)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
        at io.vertx.core.impl.ContextImpl$$Lambda$136/2025933177.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)

```

Looks like the worker threads are being blocked here:  https://github.com/quarkusio/quarkus/blob/master/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/VertxBlockingOutput.java#L96

**Comment by pmlopes** [comment]
I think all synchronized methods in that class are to prevent threading issues in `wrapInTCCLSwitch` as the current classloader is switched temporarily. If we synchronize only that switch, I guess we could remove all the other synchronized blocks. Then if you don't use any custom class loader, then you would never be inside a synchronized block at all.

**Comment by stuartwdouglas** [comment]
Why would a TCCL switch need to be synchronized? by definition it only involves a single thread.

**Comment by gsmet** [comment]
Btw, it will probably be a stupid question but why is `StaticHandlerImpl` involved at all? Does it have precedence over the paths handled by RESTEasy?

**Comment by geoand** [comment]
Maybe because the project contains a `META-INF/resources/index.html` file?

**Comment by gsmet** [comment]
But you are only targeting the JAX-RS resource in your benchmark, aren't you?

**Comment by geoand** [comment]
> But you are only targeting the JAX-RS resource in your benchmark, aren't you?

Yup, the benchmark only hit the REST endpoint. Perhaps it's some internal thing, I didn't dig deeper but I certainly could

**Comment by pmlopes** [comment]
I've isolated the locks into 2 areas, TCCL switch and stats updates for filesystem tuning. If it's guaranteed that the TCCL is safe I can even drop that one, and locks are only active while the handle is tuning itself.

**Comment by geoand** [comment]
I tried the reproducer with Vert.x 3.8.1 (what Quarkus currently uses) + the synchronized patch and I'm still seeing issues, however I think they might be different this time since I am seeing:

```
vert.x-eventloop-thread-1  Blocked CPU usage on sample: 56ms
  io.vertx.core.file.impl.FileResolver.resolveFile(String) FileResolver.java:118
  io.vertx.core.impl.VertxImpl.resolveFile(String) VertxImpl.java:803
  io.vertx.core.file.impl.FileSystemImpl$20.<init>(FileSystemImpl, Handler, String) FileSystemImpl.java:933
  io.vertx.core.file.impl.FileSystemImpl.existsInternal(String, Handler) FileSystemImpl.java:932
  io.vertx.core.file.impl.FileSystemImpl.exists(String, Handler) FileSystemImpl.java:340
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$isFileExisting$2(FileSystem, String, Handler) StaticHandlerImpl.java:273
  io.vertx.ext.web.handler.impl.StaticHandlerImpl$$Lambda$240.call()
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.wrapInTCCLSwitch(Callable) StaticHandlerImpl.java:254
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.isFileExisting(RoutingContext, String, Handler) StaticHandlerImpl.java:273
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.sendStatic(RoutingContext, String) StaticHandlerImpl.java:181
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.handle(RoutingContext) StaticHandlerImpl.java:150
  io.vertx.ext.web.handler.impl.StaticHandlerImpl.handle(Object) StaticHandlerImpl.java:57
  io.vertx.ext.web.impl.RouteImpl.handleContext(RoutingContextImplBase) RouteImpl.java:231
  io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext() RoutingContextImplBase.java:121
  io.vertx.ext.web.impl.RoutingContextImpl.next() RoutingContextImpl.java:133
  io.vertx.ext.web.impl.RouterImpl.handle(HttpServerRequest) RouterImpl.java:80
  io.vertx.ext.web.impl.RouterImpl.handle(Object) RouterImpl.java:41
  io.quarkus.resteasy.runtime.standalone.ResteasyStandaloneRecorder$1.handle(HttpServerRequest) ResteasyStandaloneRecorder.java:40
  io.quarkus.resteasy.runtime.standalone.ResteasyStandaloneRecorder$1.handle(Object) ResteasyStandaloneRecorder.java:37
  io.quarkus.vertx.http.runtime.VertxHttpRecorder$2.handle(RoutingContext) VertxHttpRecorder.java:165
  io.quarkus.vertx.http.runtime.VertxHttpRecorder$2.handle(Object) VertxHttpRecorder.java:162
  io.vertx.ext.web.impl.RouteImpl.handleContext(RoutingContextImplBase) RouteImpl.java:231
  io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext() RoutingContextImplBase.java:121
  io.vertx.ext.web.impl.RoutingContextImpl.next() RoutingContextImpl.java:133
  io.vertx.ext.web.impl.RouterImpl.handle(HttpServerRequest) RouterImpl.java:80
  io.vertx.ext.web.impl.RouterImpl.handle(Object) RouterImpl.java:41
  io.vertx.core.http.impl.WebSocketRequestHandler.handle(HttpServerRequest) WebSocketRequestHandler.java:50
  io.vertx.core.http.impl.WebSocketRequestHandler.handle(Object) WebSocketRequestHandler.java:32
  io.vertx.core.http.impl.HttpServerRequestImpl.handleBegin(Handler) HttpServerRequestImpl.java:151
  io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Object) Http1xServerConnection.java:135
  io.vertx.core.net.impl.VertxHandler$$Lambda$156.handle(Object)
  io.vertx.core.impl.ContextImpl.executeTask(Object, Handler) ContextImpl.java:369
  io.vertx.core.impl.EventLoopContext.execute(Object, Handler) EventLoopContext.java:43
  io.vertx.core.impl.ContextImpl.executeFromIO(Object, Handler) ContextImpl.java:232
  io.vertx.core.net.impl.VertxHandler.channelRead(ChannelHandlerContext, Object) VertxHandler.java:173
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Object) AbstractChannelHandlerContext.java:374
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) AbstractChannelHandlerContext.java:360
  io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(Object) AbstractChannelHandlerContext.java:352
  io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ChannelHandlerContext, CodecOutputList, int) ByteToMessageDecoder.java:328
  io.netty.handler.codec.ByteToMessageDecoder.channelRead(ChannelHandlerContext, Object) ByteToMessageDecoder.java:302
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Object) AbstractChannelHandlerContext.java:374
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) AbstractChannelHandlerContext.java:360
  io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(Object) AbstractChannelHandlerContext.java:352
  io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(ChannelHandlerContext, Object) DefaultChannelPipeline.java:1421
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(Object) AbstractChannelHandlerContext.java:374
  io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext, Object) AbstractChannelHandlerContext.java:360
  io.netty.channel.DefaultChannelPipeline.fireChannelRead(Object) DefaultChannelPipeline.java:930
  io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read() AbstractNioByteChannel.java:163
  io.netty.channel.nio.NioEventLoop.processSelectedKey(SelectionKey, AbstractNioChannel) NioEventLoop.java:697
  io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized() NioEventLoop.java:632
  io.netty.channel.nio.NioEventLoop.processSelectedKeys() NioEventLoop.java:549
  io.netty.channel.nio.NioEventLoop.run() NioEventLoop.java:511
  io.netty.util.concurrent.SingleThreadEventExecutor$5.run() SingleThreadEventExecutor.java:918
  io.netty.util.internal.ThreadExecutorMap$2.run() ThreadExecutorMap.java:74
  io.netty.util.concurrent.FastThreadLocalRunnable.run() FastThreadLocalRunnable.java:30
  java.lang.Thread.run() Thread.java:748
```

**Comment by pmlopes** [comment]
That's something else. A request is trying to load a file, the static handler isn't blocking anymore but the file resolver is locking trying to find the file from the classpath and copy to the cache. As the file doesn't exist, this will lock for every call.

**Comment by geoand** [comment]
I removed the stacktrace I had earlier since I think I read the profiling data incorrectly

**Comment by pmlopes** [comment]
The static handler should be at the end of the chain to avoid hitting the file system all the time.

**Comment by geoand** [comment]
I think @patriot1burke mentioned that currently it Quarkus it's the first handler

**Comment by geoand** [comment]
With patch, the application does report this:

```
Sep 25, 2019 12:10:25 AM io.vertx.core.impl.BlockedThreadChecker
WARNING: Thread Thread[vert.x-worker-thread-2,5,main]=Thread[vert.x-worker-thread-2,5,main] has been blocked for 115047 ms, time limit is 60000 ms
io.vertx.core.VertxException: Thread blocked
	at java.lang.Object.$$YJP$$wait(Native Method)
	at java.lang.Object.wait(Object.java)
	at java.lang.Object.wait(Object.java:502)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.awaitWriteable(VertxBlockingOutput.java:96)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.write(VertxBlockingOutput.java:63)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.writeBlocking(VertxHttpResponse.java:179)
	at io.quarkus.resteasy.runtime.standalone.VertxOutputStream.close(VertxOutputStream.java:118)
	at org.jboss.resteasy.util.CommitHeaderOutputStream.close(CommitHeaderOutputStream.java:87)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.finish(VertxHttpResponse.java:136)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:109)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:64)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:55)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$$Lambda$150/1085467586.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
	at io.vertx.core.impl.ContextImpl$$Lambda$147/637961193.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by pmlopes** [comment]
This Trace is different can you figure out which lambda is that? That calls an execute blocking?

**Comment by geoand** [comment]
This is what the application itself reports.

The previous one I had posted if what I got from profiling the application and looking at the threads.

**Comment by geoand** [comment]
I can't really tell which lambda it is TBH. Sorry I couldn't help more. I'm calling it a night :)

**Comment by geoand** [comment]
Since I am really don't know much about Vert.x I think someone with more Quarkus + Vert.x knowledge should take over. 
I hope I haven't led people down a rabbit hole...

**Comment by geoand** [comment]
Last thing for today - here is what I get from jstack with the patch applied when the application stops being able to serve requests:

```
"vert.x-worker-thread-19" #71 prio=5 os_prio=0 tid=0x00007f759400f000 nid=0x72f3 in Object.wait() [0x00007f7610bf7000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000005d4365b70> (a io.vertx.core.http.impl.Http1xServerConnection)
	at java.lang.Object.wait(Object.java:502)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.awaitWriteable(VertxBlockingOutput.java:96)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.write(VertxBlockingOutput.java:63)
	- locked <0x00000005d4365b70> (a io.vertx.core.http.impl.Http1xServerConnection)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.writeBlocking(VertxHttpResponse.java:179)
	at io.quarkus.resteasy.runtime.standalone.VertxOutputStream.close(VertxOutputStream.java:118)
	at org.jboss.resteasy.util.CommitHeaderOutputStream.close(CommitHeaderOutputStream.java:87)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.finish(VertxHttpResponse.java:136)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:109)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:64)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:55)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$$Lambda$204/1906229739.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
	at io.vertx.core.impl.ContextImpl$$Lambda$199/694340747.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

   Locked ownable synchronizers:
	- <0x00000005cf855798> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"vert.x-worker-thread-18" #70 prio=5 os_prio=0 tid=0x00007f75a000c000 nid=0x72f2 in Object.wait() [0x00007f7610cf8000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000005d43661c8> (a io.vertx.core.http.impl.Http1xServerConnection)
	at java.lang.Object.wait(Object.java:502)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.awaitWriteable(VertxBlockingOutput.java:96)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.write(VertxBlockingOutput.java:63)
	- locked <0x00000005d43661c8> (a io.vertx.core.http.impl.Http1xServerConnection)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.writeBlocking(VertxHttpResponse.java:179)
	at io.quarkus.resteasy.runtime.standalone.VertxOutputStream.close(VertxOutputStream.java:118)
	at org.jboss.resteasy.util.CommitHeaderOutputStream.close(CommitHeaderOutputStream.java:87)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.finish(VertxHttpResponse.java:136)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:109)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:64)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:55)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$$Lambda$204/1906229739.handle(Unknown Source)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
	at io.vertx.core.impl.ContextImpl$$Lambda$199/694340747.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

   Locked ownable synchronizers:
	- <0x00000005cf4050b0> (a java.util.concurrent.ThreadPoolExecutor$Worker)
```

**Comment by geoand** [comment]
@pmlopes the very first stack trace I added when I opened the issue may or may not have been misleading. I don't really know...

**Comment by johnaohara** [comment]
I thought I commented on this yesterday, but the worker threads being blocked at the end of the run is being caused by;

https://github.com/quarkusio/quarkus/blob/master/extensions/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/standalone/VertxBlockingOutput.java#L96

**Comment by johnaohara** [comment]
```
"vert.x-eventloop-thread-7" #23 prio=5 os_prio=0 cpu=177.33ms elapsed=61.69s tid=0x00007f4320aab000 nid=0x3238 waiting for monitor entry  [0x00007f42f0d88000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at io.vertx.ext.web.handler.impl.StaticHandlerImpl.isFileExisting(StaticHandlerImpl.java:275)
```
is a separate issue

**Comment by pmlopes** [comment]
Ok, so this seems to tell us that the fix upstream should be backported to 3.8.2 as it "solves" the locking in vertx-web `isFileExisting`

As @geoand now doesn't see that lock but another one at `VertxBlockingOutput`

**Comment by gsmet** [comment]
@geoand what's the status of this one? Is it fixed?

**Comment by geoand** [comment]
Yes, it was fixed in #4191
